flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/MatchCodeGenerator.scala

```diff
@@ -0,0 +1,320 @@
+        val inputTypeTerm =
+          s"java.util.Map<String, java.util.List<${boxedTypeTermForTypeInfo(input)}>>"
+
+        (baseClass,
+          s"void flatSelect($inputTypeTerm $input1Term, $collectorTypeTerm $collectorTerm)",
+          List(),
+          true)
+      } else {
+        throw new CodeGenException("Unsupported Function.")
+      }
+
+    val extendsKeyword = if (isInterface) "implements" else "extends"
+    val funcCode = j"""
+      |public class $funcName $extendsKeyword ${functionClass.getCanonicalName} {
+      |
+      |  ${reuseMemberCode()}
+      |
+      |  public $funcName() throws Exception {
+      |    ${reuseInitCode()}
+      |  }
+      |
+      |  @Override
+      |  public $signature throws Exception {
+      |    ${inputStatements.mkString("\n")}
+      |    ${reusePerRecordCode()}
+      |    ${reuseInputUnboxingCode()}
+      |    $bodyCode
+      |  }
+      |}
+    """.stripMargin
+
+    GeneratedFunction(funcName, returnType, funcCode)
+  }
+
+  def generatePartitionKeyAccess(
+    partitionKey: RexInputRef
+  ) : GeneratedExpression = {
+    // extracts partition keys from any element of the match
+    val eventNameTerm = newName("event")
+    val eventTypeTerm = boxedTypeTermForTypeInfo(input)
+
+    val keyAccess = generateFieldAccess(input, eventNameTerm, partitionKey.getIndex)
+
+    val keyCode = s"""
+       |$eventTypeTerm $eventNameTerm = null;
+       |for (java.util.Map.Entry entry : $input1Term.entrySet()) {
+       |  java.util.List<$eventTypeTerm> value = (java.util.List<$eventTypeTerm>) entry.getValue();
+       |  if (value != null && value.size() > 0) {
+       |    $eventNameTerm = ($eventTypeTerm) value.get(0);
+       |    break;
+       |  }
+       |}
+       |
+       |${keyAccess.code}
+       """.stripMargin
+
+    keyAccess.copy(code = keyCode)
+  }
+
+  def generateOneRowPerMatchExpression(
+    partitionKeys: util.List[RexNode],
+    measures: util.Map[String, RexNode],
+    returnType: RowSchema
+  ): GeneratedExpression = {
+    // For "ONE ROW PER MATCH", the output columns include:
+    // 1) the partition columns;
+    // 2) the columns defined in the measures clause.
+    val resultExprs =
+      partitionKeys.asScala.map { case inputRef: RexInputRef =>
+        generatePartitionKeyAccess(inputRef)
+      } ++ returnType.fieldNames.filter(measures.containsKey(_)).map { fieldName =>
+        generateExpression(measures.get(fieldName))
+      }
+
+    generateResultExpression(
+      resultExprs,
+      returnType.typeInfo,
+      returnType.fieldNames)
+  }
+
+  override def visitCall(call: RexCall): GeneratedExpression = {
+    call.getOperator match {
+      case PREV | NEXT =>
+        val countLiteral = call.operands.get(1).asInstanceOf[RexLiteral]
+        val count = countLiteral.getValueAs(classOf[JLong])
+        if (count != 0) {
+          throw new TableException("Flink does not support physical offsets within partition.")
```

Comment:
No we do not support physical offsets at all. Physical offset index elements independently from which rows were matched, e.g. you can index to before the start of a match. We do not have access to such elements.

The `if` clause here is only because calcite translates every PRV into `PREV` call.

Reply:
OK, make sense.